Begin implementing back/forward cache for site isolation

achristensen07 opened the pull request:
#### 00dcbd8c70b8cb1baa57c6ba0ccdd5a15819887f
<pre>
Begin implementing back/forward cache for site isolation
<a href="https://bugs.webkit.org/show_bug.cgi?id=265561">https://bugs.webkit.org/show_bug.cgi?id=265561</a>
<a href="https://rdar.apple.com/118965412">rdar://118965412</a>

Reviewed by Pascoe.

My approach before was to terminate the iframe processes when we navigate away,
which seems to match other browsers but doesn&apos;t match our existing behavior of
being able to quickly navigate back to a fully functional page.  We need to keep
the processes alive and make CachedFrame able to handle RemoteFrames in order to
keep that speed.  If we get a memory pressure warning, we will remove back/forward
cache entries, but if we have the memory let&apos;s keep the existing behavior.

This is the first of at least two PRs to get back/forward navigation working with
site isolation.  I make CachedFrame able to handle RemoteFrames, I made
HistoryController a non-nested class to be able to be owned by RemoteFrames (but
I didn&apos;t change that ownership yet), and I swap RemotePageProxy state to a
SuspendedPageProxy when suspending.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::attachToCachedFrame):
* Source/WebCore/history/CachedFrame.cpp:
(WebCore::CachedFrameBase::CachedFrameBase):
(WebCore::CachedFrameBase::restore):
(WebCore::CachedFrame::CachedFrame):
(WebCore::CachedFrame::open):
(WebCore::CachedFrame::destroy):
* Source/WebCore/history/CachedFrame.h:
(WebCore::CachedFrameBase::view const):
* Source/WebCore/history/CachedPage.cpp:
(WebCore::CachedPage::CachedPage):
(WebCore::firePageShowEvent):
(WebCore::CachedPage::restore):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::open):
* Source/WebKit/UIProcess/SuspendedPageProxy.cpp:
(WebKit::SuspendedPageProxy::SuspendedPageProxy):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::removeRemotePagesForSuspension): Deleted.
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/271590@main">https://commits.webkit.org/271590@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b4b3d1e8edc08d60e7d1f317883208b56fc47a0f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28895 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7536 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30251 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31509 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26342 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29467 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9663 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4884 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26384 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29167 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6230 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24813 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5432 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5563 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25830 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/32846 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26424 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26268 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31810 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5542 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3707 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29591 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7174 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6020 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3723 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/6063 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

pascoej commented on Source/WebCore/history/CachedFrame.cpp:
```diff
@@ -266,7 +283,8 @@ void CachedFrame::destroy()
 
-    LocalFrame::clearTimers(m_view.get(), m_document.get());
+    if (RefPtr localFrameView = dynamicDowncast<LocalFrameView>(m_view.get()))
+        LocalFrame::clearTimers(localFrameView.get(), m_document.get());
```
Here I think you might null dereference document in LocalFrame::clearTimers when the localframe doesn't have one.

pascoej commented on Source/WebCore/history/CachedPage.cpp:
```diff
@@ -144,10 +141,11 @@ void CachedPage::restore(Page& page)
 
         bool hadProhibitsScrolling = false;
-        auto* frameView = localMainFrame->view();
+        auto* frameView = localMainFrame ? localMainFrame->view() : nullptr;
```
The only function used below is in ScrollView, which is in all FrameViews local or remote. Is there a reason we can't use `auto* frameView = mainFrame.view();` on line 148 instead?

pascoej commented on Source/WebCore/history/CachedPage.cpp:
```diff
@@ -172,14 +173,16 @@ void CachedPage::restore(Page& page)
 
     if (m_needsUpdateContentsSize) {
-        if (auto* frameView = localMainFrame->view())
+        if (auto* frameView = dynamicDowncast<LocalFrameView>(mainFrame.virtualView()))
```
ditto here, though I think in the remote case it will case an empty function defined in ScrollableArea.h (likely the case above doesn't do much of anything either, but it feels wrong not calling in it since it's in the hierarchy.)

pascoej replied:
`    virtual void updateContentsSize() { }
`

pascoej commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -2448,11 +2448,13 @@ void FrameLoader::open(CachedFrameBase& cachedFrame)
     Ref frame = m_frame.get();
     std::optional<IntRect> previousViewFrameRect = frame->view() ?  frame->view()->frameRect() : std::optional<IntRect>(std::nullopt);
-    frame->setView(view.copyRef());
+    if (RefPtr localView = dynamicDowncast<LocalFrameView>(view.get()))
```
Is it okay to set a view here even if it's remote?

achristensen07 replied:
not currently.  We also need to make sure a LocalFrame only has a LocalFrameView and a RemoteFrame only has a RemoteFrameView.

achristensen07 committed:
Begin implementing back/forward cache for site isolation
https://bugs.webkit.org/show_bug.cgi?id=265561
rdar://118965412

Reviewed by Pascoe.

My approach before was to terminate the iframe processes when we navigate away,
which seems to match other browsers but doesn't match our existing behavior of
being able to quickly navigate back to a fully functional page.  We need to keep
the processes alive and make CachedFrame able to handle RemoteFrames in order to
keep that speed.  If we get a memory pressure warning, we will remove back/forward
cache entries, but if we have the memory let's keep the existing behavior.

This is the first of at least two PRs to get back/forward navigation working with
site isolation.  I make CachedFrame able to handle RemoteFrames, I made
HistoryController a non-nested class to be able to be owned by RemoteFrames (but
I didn't change that ownership yet), and I swap RemotePageProxy state to a
SuspendedPageProxy when suspending.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::attachToCachedFrame):
* Source/WebCore/history/CachedFrame.cpp:
(WebCore::CachedFrameBase::CachedFrameBase):
(WebCore::CachedFrameBase::restore):
(WebCore::CachedFrame::CachedFrame):
(WebCore::CachedFrame::open):
(WebCore::CachedFrame::destroy):
* Source/WebCore/history/CachedFrame.h:
(WebCore::CachedFrameBase::view const):
* Source/WebCore/history/CachedPage.cpp:
(WebCore::CachedPage::CachedPage):
(WebCore::firePageShowEvent):
(WebCore::CachedPage::restore):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::open):
* Source/WebKit/UIProcess/SuspendedPageProxy.cpp:
(WebKit::SuspendedPageProxy::SuspendedPageProxy):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::removeRemotePagesForSuspension): Deleted.
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/271590@main
